consensus/src/dag/payload/manager.rs

```diff
@@ -0,0 +1,160 @@
+        node: &CertifiedNode,
+    ) -> anyhow::Result<BoxFuture<Result<Vec<SignedTransaction>, oneshot::error::RecvError>>> {
+        debug!("retrieving payload for node {}", node.id());
+        let (tx, rx) = oneshot::channel();
+        let DagPayload::Decoupled(info) = node.payload() else {
+            unreachable!("payload manager is only for decouple DAG payload")
+        };
+        self.waiters
+            .entry(*info.digest())
+            .or_insert_with(|| BoundedVecDeque::new(1))
+            .deref_mut()
+            .push_back(tx);
+        match self.payload_store.get(info.id(), info.digest()) {
+            Ok(payload) => {
+                let Payload::DirectMempool(txns) = payload.payload() else {
+                    unreachable!("other payloads are not supported");
+                };
+                debug!("payload available {}", payload.id());
+                if let Some(tx) = self
+                    .waiters
+                    .remove(info.digest())
+                    .expect("must exist")
+                    .1
+                    .pop_front()
+                {
+                    tx.send(txns.clone()).ok();
+                }
+                Ok(async move { rx.await }.boxed())
+            },
+            Err(DagPayloadStoreError::Missing(_)) => {
+                debug!("payload missing {}", info.id());
+                let responders = node.parents_metadata().map(|m| *m.author()).collect();
+                let request_rx = self.requester.request(info.clone(), responders)?;
+                let me = self.clone();
+                let fut = async move {
+                    let node_payload = request_rx.await?;
+                    let Payload::DirectMempool(txns) = node_payload.payload() else {
+                        unreachable!("other payloads are not supported");
+                    };
+                    if let Some(tx) = me
+                        .waiters
+                        .remove(info.digest())
+                        .expect("must exist")
+                        .1
+                        .pop_front()
+                    {
+                        tx.send(txns.clone()).ok();
+                    }
+                    rx.await
+                };
+                Ok(fut.boxed())
+            },
+            Err(_) => {
+                error!("unable to send request fetch {}", info.id());
+                bail!("error fetching");
+            },
+        }
+    }
+
+    fn prefetch_payload(&self, node: &CertifiedNode) {
+        let DagPayload::Decoupled(info) = node.payload() else {
+            unreachable!("payload manager is only for decouple DAG payload")
+        };
+        match self.payload_store.get(info.id(), info.digest()) {
+            Ok(_) => {},
+            Err(DagPayloadStoreError::Missing(_)) => {
+                debug!("prefetch payload missing {}", node.id());
+                let responders = node.parents_metadata().map(|m| *m.author()).collect();
```

Comment:
Again, why parents?

Reply:
this should be the signers of the certificate?